### PR TITLE
Added sign out functionality

### PR DESCRIPTION
**Pull request by toddsouthenbentley**

### CODE REVIEW

File: mobile-sdk/src/main/java/com/github/itwin/mobilesdk/ITMOIDCAuthorizationClient.kt [hunk 1]
```diff
@@ -194,6 +198,65 @@ open class ITMOIDCAuthorizationClient(private val itmApplication: ITMApplication
...
+    private suspend fun revokeToken(token: String, revokeUrl: URL, authorization: String) {
+        withContext(Dispatchers.IO) {
+            val connection = revokeUrl.openConnection() as HttpURLConnection
+            try {
+                val bodyBytes = "token=$token".toByteArray()
+                connection.requestMethod = "POST"
+                connection.setRequestProperty("Authorization", "Basic $authorization")
+                connection.setRequestProperty("Content-Type", "application/x-www-form-urlencoded")
+                connection.setRequestProperty("Content-Length", "${bodyBytes.size}")
+                connection.useCaches = false
```
Comment (tcobbs-bentley) [inline mobile-sdk/src/main/java/com/github/itwin/mobilesdk/ITMOIDCAuthorizationClient.kt:211]: FYI, I don't know if the iOS code not doing the equivalent of this is a problem, but I'll update it to do so.

File: mobile-sdk/src/main/java/com/github/itwin/mobilesdk/ITMOIDCAuthorizationClient.kt [hunk 2]
```diff
@@ -194,6 +198,65 @@ open class ITMOIDCAuthorizationClient(private val itmApplication: ITMApplication
...
+        val lastTokenResponse = currAuthState.lastTokenResponse
+            ?: throw Error("Have a cached authState but no lastTokenResponse")
+        val revokeURLString = currAuthState.authorizationServiceConfiguration?.discoveryDoc?.docJson?.optString("revocation_endpoint")
+            ?: throw Error("Revocation endpoint not found")
+        val revokeURL = URL(revokeURLString)
+        if (!revokeURL.protocol.equals("https", true)) {
+            throw Error("Invalid protocol (${revokeURL.protocol}) in revocation endpoint URL")
+        }
+        val authorization = Base64.getEncoder().encodeToString("${authSettings.clientId}:".toByteArray())
+        val tokens = setOfNotNull(lastTokenResponse.idToken, lastTokenResponse.accessToken, lastTokenResponse.refreshToken)
```
Comment (tcobbs-bentley) [inline mobile-sdk/src/main/java/com/github/itwin/mobilesdk/ITMOIDCAuthorizationClient.kt:237]: The logic is different here from iOS. We need to decide which has the correct logic and use it for both. I will discuss with you.

File: mobile-sdk/src/main/java/com/github/itwin/mobilesdk/ITMOIDCAuthorizationClient.kt [hunk 3]
```diff
@@ -194,6 +198,65 @@ open class ITMOIDCAuthorizationClient(private val itmApplication: ITMApplication
...
+                ex.message?.let { errList.add(it) }
+            }
+        }
+        if (errList.isNotEmpty()) {
+            throw Error (errList.toTypedArray().joinToString("\n"))
+        }
+    }
+
+    @Suppress("unused")
+    fun signOut() {
```
Comment (tcobbs-bentley) [inline mobile-sdk/src/main/java/com/github/itwin/mobilesdk/ITMOIDCAuthorizationClient.kt:252]: I feel that this should be a suspend fun.

File: mobile-sdk/src/main/java/com/github/itwin/mobilesdk/ITMOIDCAuthorizationClient.kt (hunk 2 above)
Comment (tcobbs-bentley) [inline mobile-sdk/src/main/java/com/github/itwin/mobilesdk/ITMOIDCAuthorizationClient.kt:237]: I think that you should get these values from authState, not lastTokenResponse.

File: mobile-sdk/src/main/java/com/github/itwin/mobilesdk/ITMOIDCAuthorizationClient.kt [hunk 4]
```diff
@@ -194,6 +198,65 @@ open class ITMOIDCAuthorizationClient(private val itmApplication: ITMApplication
...
+    }
+
+    private suspend fun revokeTokens() {
+        val currAuthState = authState ?: return
+        val lastTokenResponse = currAuthState.lastTokenResponse
+            ?: throw Error("Have a cached authState but no lastTokenResponse")
+        val revokeURLString = currAuthState.authorizationServiceConfiguration?.discoveryDoc?.docJson?.optString("revocation_endpoint")
+            ?: throw Error("Revocation endpoint not found")
+        val revokeURL = URL(revokeURLString)
+        if (!revokeURL.protocol.equals("https", true)) {
```
Comment (tcobbs-bentley) [inline mobile-sdk/src/main/java/com/github/itwin/mobilesdk/ITMOIDCAuthorizationClient.kt:233]: I feel you should only do this if tokens is non-empty.

File: mobile-sdk/src/main/java/com/github/itwin/mobilesdk/ITMOIDCAuthorizationClient.kt (hunk 2 above)
Comment (toddsouthenbentley) [reply mobile-sdk/src/main/java/com/github/itwin/mobilesdk/ITMOIDCAuthorizationClient.kt]: Fixed, logic should be identical (along with error messages)

File: mobile-sdk/src/main/java/com/github/itwin/mobilesdk/ITMOIDCAuthorizationClient.kt (hunk 3 above)
Comment (toddsouthenbentley) [reply mobile-sdk/src/main/java/com/github/itwin/mobilesdk/ITMOIDCAuthorizationClient.kt]: Fixed

File: mobile-sdk/src/main/java/com/github/itwin/mobilesdk/ITMOIDCAuthorizationClient.kt [hunk 5]
```diff
@@ -194,6 +198,65 @@ open class ITMOIDCAuthorizationClient(private val itmApplication: ITMApplication
...
+                }
+            } finally {
+                connection.disconnect()
+            }
+        }
+    }
+
+    private suspend fun revokeTokens() {
+        val currAuthState = authState ?: return
+        val lastTokenResponse = currAuthState.lastTokenResponse
```
[the remark this answers is not in the document]
Comment (toddsouthenbentley) [reply mobile-sdk/src/main/java/com/github/itwin/mobilesdk/ITMOIDCAuthorizationClient.kt]: Removed in favor of using the authState getters.

File: mobile-sdk/src/main/java/com/github/itwin/mobilesdk/ITMOIDCAuthorizationClient.kt (hunk 2 above)
Comment (toddsouthenbentley) [reply mobile-sdk/src/main/java/com/github/itwin/mobilesdk/ITMOIDCAuthorizationClient.kt]: Fixed

File: mobile-sdk/src/main/java/com/github/itwin/mobilesdk/ITMOIDCAuthorizationClient.kt (hunk 4 above)
Comment (toddsouthenbentley) [reply mobile-sdk/src/main/java/com/github/itwin/mobilesdk/ITMOIDCAuthorizationClient.kt]: Fixed

**Review by tcobbs-bentley — APPROVED** [review]

**Pull request merged**